### ISSUE TITLE
Queue: Case movement | Admins moving cases

### ISSUE THREAD

**Issue by laurjpeterson** [department-of-veterans-affairs/caseflow#11801]
## Context

Caseflow built critical functionality for AMA cases to be processed in time for the law’s implementation. The team focused on the most straightforward path (also known as the happy path) an appeal could take to meet this deadline, and get to a “minimum viable” state.

Caseflow is now iterating on this initial feature set to handle other paths an appeal could take.

This ticket encompasses allowing users to administratively act on behalf of other users.

## Goals for this functionality:
- Allow the Board to complete daily case movement tasks without support/help desk intervention

### Allow chosen users to act on behalf of judges

**Context:** Sometimes, judges are out on leave, and their supervisors or other users need to reassign cases (mainly JudgeDecisionReviewTasks) to other judges so the appeal isn't held up

**Stories**
- Allow chosen users to reassign judge tasks to other judges to account for judges leave, sickness, etc.

### Allow chosen users to act on behalf of DVCs

**Context:** Deputy Vice Chairmen have staff assistants that work for them. They want their staff assistants to be able to administratively act on behalf of them within Caseflow to complete certain tasks:
- Requesting cases for DVCs, so they can be assigned to judges
- Assigning those requested cases to other judges

This scenario isn't much different than the above one, because DVCs are also considered judges in VACOLS and Caseflow. But, I've separated it out because our stakeholders view them differently. 

**Stories**
- Allow chosen users to request cases and assign tasks to judges on behalf of DVCs

## NEW Implementation
Allow case movement team to perform all judge movement actions

## DEPRECIATED Implementation
<details>
  <summary>Click to see depreciated implementation</summary>

As worked out in #12175. 

### Description
JudgeTeams will be extended to support admins outside of the Judge themselves. We are creating a new way to track who is the judge and who are the attorneys.

Anyone who is Admin on a judge team would have the ability to access the JudgeQueue for that team, request cases, and assign out work. These actions would all report as performed by the Judge on the team. The scope of this work is Case Management (Request Cases, AttorneyTask Assign/Reassign, Reassign to another Judge), with any other Judge abilities (hearings prep, decision review) to follow in future work, if requested.

### Scenario Applications

Both scenarios are handled the same way.
**Allow chosen users to act on behalf of judges**
**Allow chosen users to act on behalf of DVCs**

Users can be added as an Admin to any JudgeTeams they want to act on behalf of. These users can either be attorneys on the team or not (marked appropriately on the Team page). For existing Judges we add these users to their teams as an Admin. For DVCs without a JudgeTeam, we create one with them as the Judge and their staff assistants as additional Admins. If they already had a JudgeTeam, it proceeds the same as the other case.

### Work Breakdown
Breakout of the work for this epic, based on the tech spec #12175 
1. Extend organizations_user with judge & attorney roles #12420
1. Migrate the existing Judge Teams #12421 (blocked by 12420)
1. Update existing Judge & Attorney assumptions based on judge_team admin/non-admin to use the roles #12422 (blocked by 12420)
1. Add Judge & Attorney Selectors to JudgeTeam Team Page #12433  (blocked by 12422)
1. Give JudgeTeam Admins dropdowns to select their JudgeTeam queues #12423 (blocked by 12422)
1. Update Judge actions to be available to any admins on their Judge team
   1. Case Management #12424 (blocked by 12422)
      1. Admins can Request Cases on behalf of the Judge
      1. Admins can Assign or Reassign Cases to Attorneys on behalf of the Judge
      1. Admins can Reassign Cases to another Judge on behalf of the Judge
   1. Other Abilities (Optional if requested or implemented incidental to above, but not a part of this effort)
       1. Hearing Judge Activities 
       1. Decision Review
1. Add relevant users to judge teams as admins. Update Training Documentation.
</details>

**Comment by pkarman** [comment]
Just a technical note to mention that there is prior art for this kind of "act as another user" functionality that we should consider when plotting the technical implementation.

**Comment by lomky** [comment]
Breakout of the work for this epic, based on the tech spec #12211
1. Extend organizations_user with judge & attorney roles #12420
1. Migrate the existing Judge Teams #12421
1. Update existing Judge & Attorney assumptions based on judge_team admin/non-admin to use the roles #12422
1. Give JudgeTeam Admins dropdowns to select their JudgeTeam queues #12423
1. Update Judge actions to be available to any admins on their Judge team
   1. Case Management (Request, AttorneyTask Assign/Reassign/Cancel, Reassign to another Judge) #12424
   1. Other Abilities (Optional, on request, non-blocking
       1. Hearing Judge Activities 
       1. Decision Review
1. Add relevant users to judge teams as admins. Update Training Documentation.

**Comment by laurjpeterson** [comment]
@lomky q for you about the above:

> The scope of this work is Case Management (Request Cases, AttorneyTask Assign/Reassign, Reassign to another Judge), with any other Judge abilities (hearings prep, decision review) to follow in future work, if requested.

Does "Reassign to another judge" include reassigning JudgeDecisionReviewTasks to other judges? Or just JudgeAssignTasks?

**Comment by laurjpeterson** [comment]
Another quick q to think about (and I realize it might be late). There are ~100 judges. If all 4 DVCs want the ability to act on behalf of all 100 judges, how might we assign them to be admins of all 100 judges in bulk instead of one by one?

**Comment by lomky** [comment]
> Does "Reassign to another judge" include reassigning JudgeDecisionReviewTasks to other judges? Or just JudgeAssignTasks?

 All Judge*Tasks!

> Another quick q to think about (and I realize it might be late). There are ~100 judges. If all 4 DVCs want the ability to act on behalf of all 100 judges, how might we assign them to be admins of all 100 judges in bulk instead of one by one?

We could do this in bulk for the existing teams. Then it would be up to the org to do it moving forward with new JudgeTeams. We could look at doing additional work maybe around automatically adding certain users as admins to newly created JudgeTeams, but that's out of scope of this work.

**Comment by araposo-tistatech** [comment]
Re-opening as this epic is to complete the work for moving non-blocking tasks by the Case Movement team. 

@geoffreystienblock was there a specific reason why this epic was closed?
